### PR TITLE
Add support for image and image aliases types

### PR DESCRIPTION
**Pull request by stgraber**

### CODE REVIEW

File: shared/api/image.go [hunk 1]
```diff
@@ -34,6 +34,9 @@ type ImagesPostSource struct {
 	// For type "image"
 	Fingerprint string `json:"fingerprint" yaml:"fingerprint"`
 	Secret      string `json:"secret" yaml:"secret"`
+
+	// API extension: image_types
+	ImageType string `json:"image_type" yaml:"image_type"`
```
Comment (tomponline) [inline shared/api/image.go]: @stgraber is it worth making ImageType the field name on all of the image related structs, so they are consistent with each other, as Type is taken already in this stuct. (I just fixed a bug in my code where I got confused between InstanceType and Type).

File: lxd/containers_post.go [hunk 2]
```diff
@@ -113,7 +113,7 @@ func createFromImage(d *Daemon, project string, req *api.ContainersPost) Respons
 			}
 			info, err = d.ImageDownload(
 				op, req.Source.Server, req.Source.Protocol, req.Source.Certificate,
-				req.Source.Secret, hash, true, autoUpdate, "", true, project)
+				req.Source.Secret, hash, "container", true, autoUpdate, "", true, project)
```
Comment (tomponline) [inline lxd/containers_post.go]: @stgraber you could use api.InstanceTypeContainer if you liked.

File: lxd/daemon_images.go [hunk 3]
```diff
@@ -425,6 +429,11 @@ func (d *Daemon) ImageDownload(op *operation, server string, protocol string, ce
 			}
 		}
 
+		// Compatibility with older LXD servers
+		if info.Type == "" {
+			info.Type = "container"
```
Comment (tomponline) [inline lxd/daemon_images.go]: @stgraber same as above, using api.InstanceTypeContainer would make it easier to find in the future.

File: lxd/db/images.go [hunk 4]
```diff
@@ -11,14 +11,19 @@ import (
...
-// names.
+// ImageSourceProtocol maps image source protocol codes to human-readable names.
 var ImageSourceProtocol = map[int]string{
 	0: "lxd",
 	1: "direct",
 	2: "simplestreams",
 }
 
+// ImageType maps image types to human-readable names.
+var ImageType = map[int]string{
```
Comment (tomponline) [inline lxd/db/images.go]: @stgraber could this use the instance.TypeX consts, or perhaps even move this map to `instance` package as the names and IDs are the same as we use for instance type in DB.

File: lxd/db/images.go [hunk 5]
```diff
@@ -785,6 +809,18 @@ func (c *Cluster) ImageInsert(project, fp string, fname string, sz int64, public
 		arch = 0
 	}
 
+	imageType := -1
```
Comment (tomponline) [inline lxd/db/images.go]: @stgraber if we change `instance.New()` to return a more generic "invalid type" style message, you could use that here (and would require us to add a virtual-machine entry which I will need to do anyway).

**Comment by tomponline** [comment]
@stgraber looks good, the only thing was whether we could share the constants and text to type conversion logic used for instance types (seeing as the image type is linked to the instance type), would potentially make it easier to see in the future that the 2 are interlinked.

File: shared/api/image.go (hunk 1 above)
Comment (stgraber) [reply shared/api/image.go]: api.Image.ImageType feels weird. To be fair this one ^ feels weird too but yeah, didn't have much of a choice as I can't break backward compatibility...

So effectively treating `Type` in this context as `SourceType` and using `ImageType` as `Type`.

**Comment by stgraber** [comment]
Yep, good point about just using the instance type for the internal stuff as they do line up.

File: shared/api/image.go (hunk 1 above)
Comment (tomponline) [reply shared/api/image.go]: @stgraber yes the stuttering isn't great either.

**Comment by stgraber** [comment]
@tomponline all feedback should have been addressed

**Comment by tomponline** [comment]
Looks good thanks

**Pull request merged**